### DISCUSSION TITLE
Readonly Smart Contract Calls

### DISCUSSION THREAD

**Discussion by damip** [massalabs/massa#1970] [Smart contract engine]
# Rationale

When using Massa in a Web3 context for example, one should be able to perform read-only Smart Contract calls.

See: https://ethereum.stackexchange.com/questions/765/what-is-the-difference-between-a-transaction-and-a-call/770

# Massa implementation

## API

Add a "sc_readonly_call" API endpoint 

Parameters:
* max_gas: u64  // max gas allowed for the readonly run
* simulated_gas_price: Amount  // simulated gas price to expose to the smart contract context
* simulated_caller: Address // pretend this address is executing the SC
* bytecode: `Vec<u8>`  // bytecode to execute

Return value:
* executed_at: Slot  // slot at which the execution occurred
* result:
  * (optional) error: Error 
  * (optional) output_events: `Vec<SCOutputEvent>`  // output events generated during execution

 ## Operation

* when the sc_readonly_call is called, the bytecode's main() function will be called with the following execution context:
  * the execution will be done from the point of view of the latest slot at the current timestamp (see VM slot filler)
  * set the call stack to simulated_caller_address
  * set max_gas to its chosen value
  * set gas_price to simulated_gas_price
  * TODO: block ? maybe just assume a miss

* during the call, everything happens as with a normal ExecuteSC call, but when the call finishes, its effects are rollbacked (like when a SC execution fails)

**Comment by AureliaDolo** [comment]
How do you manage the "even if there is currently no block there" part ?

**Reply by damip** [reply]
I think we should slightly modify the current implementation of Execution in order to assume misses for all slots between the latest received block and the slot at UTime::now(). What do you think ?

**Reply by damip** [reply]
I implemented the slot filler https://github.com/massalabs/massa/pull/1981  so that's done

**Comment by gterzian** [comment]
Opened a partner issue at https://github.com/massalabs/massa/issues/1977

**Comment by damip** [comment]
updated with Greg's comments

**Comment by gterzian** [comment]
@damip Should we publish the spec in a markdown spec so the code can link to it? I guess we can also link to here...

**Reply by damip** [reply]
yes, feel free to copy the markdown into a wiki page like we did elsewhere and edit it to match the step numbers in the code

```
# Rationale

When using Massa in a Web3 context for example, one should be able to perform read-only Smart Contract calls.

See: https://ethereum.stackexchange.com/questions/765/what-is-the-difference-between-a-transaction-and-a-call/770

# Massa implementation

## API

Add a "sc_readonly_call" API endpoint 

Parameters:
* max_gas: u64  // max gas allowed for the readonly run
* simulated_gas_price: Amount  // simulated gas price to expose to the smart contract context
* simulated_caller: Address // pretend this address is executing the SC
* bytecode: `Vec<u8>`  // bytecode to execute

Return value:
* executed_at: Slot  // slot at which the execution occurred
* result:
  * (optional) error: Error 
  * (optional) output_events: `Vec<SCOutputEvent>`  // output events generated during execution

 ## Operation

* when the sc_readonly_call is called, the bytecode's main() function will be called with the following execution context:
  * the execution will be done from the point of view of the latest slot at the current timestamp (see VM slot filler)
  * set the call stack to simulated_caller_address
  * set max_gas to its chosen value
  * set gas_price to simulated_gas_price
  * TODO: block ? maybe just assume a miss

* during the call, everything happens as with a normal ExecuteSC call, but when the call finishes, its effects are rollbacked (like when a SC execution fails)


```
